### PR TITLE
feat: most booru sites apikey support

### PR DESCRIPTION
**Pull request by MaikoTan**

### CODE REVIEW

**Comment by MaikoTan** [comment]
@Seidko @Lipraty Done all sources currently. Please review.

Note: safebooru / lolicon / local / moehu don't have an API token or authentication method.
Note 2: e621 is not tested
Note 3: yande.re seems not reading token, it works even you give a random value.

File: packages/sankaku/src/index.ts [hunk 1]
```diff
@@ -8,17 +9,36 @@ class SankakuComplexImageSource extends ImageSource<SankakuComplexImageSource.Co
 
   constructor(ctx: Context, config: SankakuComplexImageSource.Config) {
     super(ctx, config)
+    this.http = this.http.extend({
+      headers: {
+        'User-Agent':
+          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
+      },
```
Comment (Lipraty) [inline packages/sankaku/src/index.ts:16]:
```suggestion
      headers: {
        'User-Agent': config.userAgent,
      },
```

File: packages/sankaku/src/index.ts [hunk 2]
```diff
@@ -32,17 +52,51 @@ class SankakuComplexImageSource extends ImageSource<SankakuComplexImageSource.Co
...
+  export interface KeyPair {
+    login?: string
+    password?: string
+    tokenType?: string
+    accessToken?: string
+  }
+
   export interface Config extends ImageSource.Config {
-    endpoint: string
+    keyPairs: KeyPair[]
```
Comment (Lipraty) [inline packages/sankaku/src/index.ts:86]:
```suggestion
    keyPairs: KeyPair[]
    userAgent: string
```

File: packages/sankaku/src/index.ts [hunk 3]
```diff
@@ -32,17 +52,51 @@ class SankakuComplexImageSource extends ImageSource<SankakuComplexImageSource.Co
...
     Schema.object({
-      endpoint: Schema.string().description('SankakuComplex 的 URL。').default('https://capi-v2.sankakucomplex.com/'),
+      keyPairs: Schema.array(
+        Schema.object({
+          login: Schema.string().required().description('SankakuComplex 用户名'),
+          password: Schema.string().required().role('secret').description('SankakuComplex 密码'),
+          tokenType: Schema.string().hidden().default('Bearer').description('SankakuComplex 访问令牌类型'),
+          accessToken: Schema.string().hidden().description('SankakuComplex 访问令牌'),
+        }).description('SankakuComplex 的登录凭证'),
+      ),
```
Comment (Lipraty) [inline packages/sankaku/src/index.ts:99]:
```suggestion
      keyPairs: Schema.array(
        Schema.object({
          login: Schema.string().required().description('SankakuComplex 用户名'),
          password: Schema.string().required().role('secret').description('SankakuComplex 密码'),
          tokenType: Schema.string().hidden().default('Bearer').description('SankakuComplex 访问令牌类型'),
          accessToken: Schema.string().hidden().description('SankakuComplex 访问令牌'),
        }).description('SankakuComplex 的登录凭证'),
      ),
      userAgent: Schema.string()
          .description('设置请求的 User Agent。')
          .default('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'),
```

**Pull request merged**